File: src/2025-03-04-03-28-40--deepseek-chat-0304/javascript/exercises/practice/parallel-letter-frequency/parallel-letter-frequency.js
// Helper function to count letters in a single text
const countLetters = (text) => {
  const frequency = {};
  for (const char of text.toLowerCase()) {
    if (/^\p{L}$/u.test(char)) {
      frequency[char] = (frequency[char] || 0) + 1;
    }
  }
  return frequency;
};

// Merge multiple frequency maps into one
const mergeFrequencies = (frequencyMaps) => {
  const result = {};
  for (const freq of frequencyMaps) {
    for (const [char, count] of Object.entries(freq)) {
      result[char] = (result[char] || 0) + count;
    }
  }
  return result;
};

export const parallelLetterFrequency = async (texts) => {
  // Process each text concurrently
  const promises = texts.map(text => 
    Promise.resolve().then(() => countLetters(text))
  );
  
  // Wait for all counts to complete
  const frequencyMaps = await Promise.all(promises);
  
  // Merge all frequency maps
  return mergeFrequencies(frequencyMaps);
};

File: src/2025-03-04-03-28-40--deepseek-chat-0304/cpp/exercises/practice/parallel-letter-frequency/parallel_letter_frequency.cpp
#include "parallel_letter_frequency.h"
#include <mutex>

static std::mutex mutex;

namespace parallel_letter_frequency {

std::map<char, int> frequency(const std::vector<std::string>& texts) {
    std::map<char, int> frequency_map;
    
    // Parallel processing of each text
    std::for_each(std::execution::par, texts.begin(), texts.end(),
        [&](const std::string& text) {
            std::map<char, int> local_map;
            
            // Count letters in current text
            for (char ch : text) {
                if (std::isalpha(ch)) {
                    local_map[std::tolower(ch)]++;
                }
            }
            
            // Merge local counts into global map
            std::lock_guard<std::mutex> lock(mutex);
            for (const auto& [letter, count] : local_map) {
                frequency_map[letter] += count;
            }
        });
    
    return frequency_map;
}

}
